darshan-util/pydarshan/darshan/backend/cffi_backend.py

```diff
@@ -565,29 +565,31 @@ def log_get_dxt_record(log, mod_name, reads=True, writes=True, dtype='dict'):
 
     size_of = ffi.sizeof("struct dxt_file_record")
     segments = ffi.cast("struct segment_info *", buf[0] + size_of  )
-
+    arr_write = np.recarray(wcnt, dtype=[("offset", int),
+                                         ("length", int),
+                                         ("start_time", float),
+                                         ("end_time", float)])
+    arr_read = np.recarray(rcnt, dtype=[("offset", int),
+                                        ("length", int),
+                                        ("start_time", float),
+                                        ("end_time", float)])
 
     for i in range(wcnt):
-        seg = {
-            "offset": segments[i].offset,
-            "length": segments[i].length,
-            "start_time": segments[i].start_time,
-            "end_time": segments[i].end_time
-        }
-        rec['write_segments'].append(seg)
-
-
-    for i in range(rcnt):
-        i = i + wcnt
-        seg = {
-            "offset": segments[i].offset,
-            "length": segments[i].length,
-            "start_time": segments[i].start_time,
-            "end_time": segments[i].end_time
-        }
-        rec['read_segments'].append(seg)
+        arr_write[i, ...] = (segments[i].offset,
+                             segments[i].length,
+                             segments[i].start_time,
+                             segments[i].end_time)
```

Comment:
@carns @shanedsnyder @jakobluettgau I feel like there should be something much cleaner/faster here. Since we can iterate over `segments` in Python I feel like we should be able to stride over the `segment_info` buffer in some kind of organized way. Each segment appears to be fixed size, with two ints and two floats, right?

I played a bit with `np.frombuffer()` earlier today, but didn't quite get it to work. I'm not sure if this is pushing up against gh-497 already, but I don't think so, since I'm only talking about the segments in a given record, rather than all the records in a log file.

If you know off the top of your head how the C memory layout works for the `segments` from `struct segment info *` that may be helpful to know. I'm sure it can be dug up from the source of course.

Reply:
This all assumes that we can live with changing the DXT segment storage format, but I think it is pretty clear that the memory efficiency of thousands of pure Python objects isn't going to work.

Reply:
We discussed this live the other day, but belatedly for posterity, here is the definition of the binary format for the dxt data in the logs:

https://github.com/darshan-hpc/darshan/blob/main/include/darshan-dxt-log-format.h#L19

Each segment is indeed a fixed size (all 64 bit values; there shouldn't be any padding in the struct) and the segments are laid out linearly in the log.